File: example/get_sound_speed_press.cpp
//------------------------------------------------------------------------------
// © 2021-2023. Triad National Security, LLC. All rights reserved.  This
// program was produced under U.S. Government contract 89233218CNA000001
// for Los Alamos National Laboratory (LANL), which is operated by Triad
// National Security, LLC for the U.S.  Department of Energy/National
// Nuclear Security Administration. All rights in the program are
// reserved by Triad National Security, LLC, and the U.S. Department of
// Energy/National Nuclear Security Administration. The Government is
// granted for itself and others acting on its behalf a nonexclusive,
// paid-up, irrevocable worldwide license in this material to reproduce,
// prepare derivative works, distribute copies to the public, perform
// publicly and display publicly, and to permit others to do so.
//------------------------------------------------------------------------------

// C++ headers
#include <cmath>
#include <iostream>

// This contains useful tools for preventing things like divide by zero
#include <singularity-eos/base/robust_utils.hpp>
// Needed to import the eos models
#include <singularity-eos/eos/eos.hpp>
// One way of initializing models with modifiers
#include <singularity-eos/eos/eos_builder.hpp>

// For simplicity. Most things live in the singularity namespace
using namespace singularity;

// Loop through a flattened list of cells and fill pressure and sound
// speed using a given EOS. We make the call two different ways, once
// with single calls, and once with the FillEos call, which should be
// more performant.
inline void PressureSoundSpeedFromDensityEnergyDensity(double *rho, // inputs
                                                       double *uu, EOS &eos,
                                                       double *P, // outputs
                                                       double *cs, const int Ncells) {
  // Allocate lambda object, which is used for caching by the tables.
  // Lambdas can do other things, such as take extra parameters as well.
  //
  // Technically this object should be allocated outside this function
  // and passed in, so that it can be persistent. But it's here for
  // example's sake.
  //
  // For parallel execution, you need one lambda vector per thread.
  // Since we're running in serial, we'll just use one.
  const int nlambda = eos.nlambda(); // get number of elements per lambda
  std::vector<double> lambda(nlambda);

  // Loop through the cells and use the two function calls
  for (int i = 0; i < Ncells; ++i) {
    double sie = robust::ratio(uu[i], rho[i]); // convert to specific internal energy
    P[i] = eos.PressureFromDensityInternalEnergy(rho[i], sie, lambda.data());
    double bmod = eos.BulkModulusFromDensityInternalEnergy(rho[i], sie, lambda.data());
    cs[i] = std::sqrt(robust::ratio(bmod, rho[i]));
  }

  /*
   * request the output variables pressure and bulk modulus.
   * Use bitwise or to build the bitfield of quantities you want.
   * Available variables for input/output are:
   *
   * density, specific_internal_energy, pressure,
   * temperature, specific_heat, bulk_modulus
   *
   * Note that all quantities not requested as output are considered
   * inputs.  Moreover, most equations of state calls require density
   * and either temperature or energy as one of their inputs.
   * Therefore, to tell the machinery that we want to use energy as an
   * input, not temperature, we must also request temperature as an
   * output.
   */
  constexpr unsigned long output =
      (thermalqs::temperature | thermalqs::pressure | thermalqs::bulk_modulus);

  // Loop through cells and use the FillEos function call
  for (int i = 0; i < Ncells; ++i) {
    double temp, cv;
    // FillEos is very general and is capable of modifying any of the inputs,
    // so const vars cannot be passed into it. However, it is often more performant
    // than making individual function calls.
    Real sie = robust::ratio(uu[i], rho[i]); // convert to specific internal energy
    eos.FillEos(rho[i], temp, sie, P[i], cv, cs[i], output, lambda.data());
    // convert bulk modulus to cs
    cs[i] = std::sqrt(robust::ratio(cs[i], rho[i]));
  }
}

int main() {

  // Parameters for ideal gas
  constexpr double gm1 = 0.6;
  constexpr double Cv = 2;

  // First initialize the EOS the "easy" way:
  EOS eos1 = IdealGas(gm1, Cv);

  // If you're on device, you need to call
  // eos1.GetOnDevice();
  // to call the EOS on device
  // However, since we're doing everything on host here, we don't make this call

  // Make some arrays
  constexpr int N = 50;
  std::vector<double> rho(N);
  std::vector<double> uu(N);
  std::vector<double> P(N);
  std::vector<double> cs(N);

  // Here we fill the rho and uu arrays with something sensible.
  for (int i = 0; i < N; ++i) {
    rho[i] = 1 + 0.1 * std::sin(2 * M_PI * i / static_cast<double>(N));
    uu[i] = 1e-2 * rho[i];
  }

  // Call it!
  PressureSoundSpeedFromDensityEnergyDensity(rho.data(), uu.data(), eos1, P.data(),
                                             cs.data(), N);

  // And let's print out the final value just for fun
  std::cout << "The final values are:\n"
            << "rho = " << rho[N - 1] << "\n"
            << "uu  = " << uu[N - 1] << "\n"
            << "P   = " << P[N - 1] << "\n"
            << "cs  = " << cs[N - 1] << std::endl;

  // It's good practice to call Finalize() after you're done using an EOS.
  // This usually only does anything if you're on device, but for GPU-data it
  // will call the appropriate destructor.
  eos1.Finalize();

  return 0;
}

File: example/get_sound_speed_press.py
#------------------------------------------------------------------------------
# © 2021-2023. Triad National Security, LLC. All rights reserved.  This
# program was produced under U.S. Government contract 89233218CNA000001
# for Los Alamos National Laboratory (LANL), which is operated by Triad
# National Security, LLC for the U.S.  Department of Energy/National
# Nuclear Security Administration. All rights in the program are
# reserved by Triad National Security, LLC, and the U.S. Department of
# Energy/National Nuclear Security Administration. The Government is
# granted for itself and others acting on its behalf a nonexclusive,
# paid-up, irrevocable worldwide license in this material to reproduce,
# prepare derivative works, distribute copies to the public, perform
# publicly and display publicly, and to permit others to do so.
#------------------------------------------------------------------------------

from singularity_eos import IdealGas, thermalqs
import numpy as np
import math

SMALL = 1e-20 # to avoid dividing by zero

def PressureSoundSpeedFromDensityEnergyDensity(rho, uu, eos, # inputs
                                               P, cs, Ncells # outputs
                                               ):
    """
    Loop through a flattened list of cells and fill pressure and sound
    speed using a given EOS. 
    """
    # Allocate lambda object, which is used for caching by the tables.
    # Lambdas can do other things, such as take extra parameters as well.
    #
    # Technically this object should be allocated outside this function
    # and passed in, so that it can be persistent. But it's here for
    # example's sake.
    #
    # For parallel execution, you need one lambda vector per thread.
    # Since we're running in serial, we'll just use one.
    nlambda = eos.nlambda # get number of elements per lambda
    lmbda = np.zeros(nlambda, dtype=np.double)

    # Option 1: Loop through the cells and use the two function calls
    for i in range(Ncells):
      sie = uu[i] / (rho[i] + SMALL) # convert to specific internal energy
      P[i] = eos.PressureFromDensityInternalEnergy(rho[i], sie, lmbda)
      bmod = eos.BulkModulusFromDensityInternalEnergy(rho[i], sie, lmbda)
      cs[i] = math.sqrt(bmod / (rho[i] + SMALL))

    print("Results from Option 1 (loop with two calls):")
    print_results()

    # Option 2: use vector functions to work on entire arrays
    lmbda = np.zeros((Ncells,nlambda), dtype=np.double)
    sie = uu / (rho + SMALL)
    bmod = np.zeros(Ncells, dtype=np.double)
    eos.PressureFromDensityInternalEnergy(rho, sie, P, lmbda)
    eos.BulkModulusFromDensityInternalEnergy(rho, sie, bmod, lmbda)
    cs[:] = np.sqrt(bmod / (rho + SMALL))

    print("Results from Option 2 (two vector function calls):")
    print_results()
  
    # Option 3: Loop through the cells and use FillEos call
    for i in range(Ncells):
      sie = uu[i] / (rho[i] + SMALL) # convert to specific internal energy
      # FillEos is very general. All arguments are considered inputs, the remaining are outputs.
      # If this auto-detection is not what is needed, you can specify a limited number of outputs with the output bitmask.
      # Note that in this case any output not set will be returned as NaN.
      # FillEos is often more performant than making individual function calls.
      state = eos.FillEos(rho=rho[i], sie=sie, output=(thermalqs.temperature | thermalqs.pressure | thermalqs.bulk_modulus), lmbda=lmbda)
      # convert bulk modulus to cs
      cs[i] = math.sqrt(state.bulk_modulus / (state.density + SMALL))

    print("Results from Option 3 (loop with FillEos call):")
    print_results()

    # Option 4: use vector FillEos function to update entire arrays
    lmbda = np.zeros((Ncells,nlambda), dtype=np.double)
    sie = uu / (rho + SMALL)
    bmod = np.zeros(Ncells, dtype=np.double)
    temp = np.zeros(Ncells, dtype=np.double)
    cv = np.zeros(Ncells, dtype=np.double)
    eos.FillEos(rho, temp, sie, P, cv, bmod, (thermalqs.temperature | thermalqs.pressure | thermalqs.bulk_modulus), lmbda)
    cs[:] = np.sqrt(bmod / (rho + SMALL))

    print("Results from Option 4 (vector FillEos call):")
    print_results()

# Parameters for ideal gas
gm1 = 0.6
Cv = 2

# We initialize the eos two ways to show how it can be done
# First initialize the EOS the "easy" way:
eos1 = IdealGas(gm1, Cv)


# Make some arrays
N = 50
rho = np.zeros(N, dtype=np.double)
uu  = np.zeros(N, dtype=np.double)
P   = np.zeros(N, dtype=np.double)
cs  = np.zeros(N, dtype=np.double)

def print_results():
  print("rho =", rho[N - 1])
  print("uu  =", uu[N - 1])
  print("P   =", P[N - 1])
  print("cs  =", cs[N - 1])
  print()

# Here we fill the rho and uu arrays with something sensible.
for i in range(N):
    rho[i] = 1 + 0.1 * math.sin((2 * math.pi * i) / N)
    uu[i] = 1e-2 * rho[i]

# Call it!
PressureSoundSpeedFromDensityEnergyDensity(rho, uu, eos1, P, cs, N)

# And let's print out the final value just for fun
print("The final values are:")
print_results()

# It's good practice to call Finalize() after you're done using an EOS.
# This usually only does anything if you're on device, but for GPU-data it
# will call the appropriate destructor.
eos1.Finalize()
